### PR TITLE
[Feature:System] Implement Popup component in Vue

### PR DESCRIPTION
**Pull request by dorian451**
### Please check if the PR fulfills these requirements:

* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant
* [x] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
No popup component in Vue.

### What is the new behavior?
Adds a popup component in Vue to be used by future Vue pages. I tried to make it have a close an api to Popup.twig as I could, as well as appearence.

### Other information?
The component has `id` and `title` props to control the id and title of the popup, as well as `closebutton.id` and `closebutton.classes` to control the id/classes of the close button.
The rest of the "properties" used in  Popup.twig are implemented using Vue's [named slots mechanism](https://vuejs.org/guide/components/slots.html#named-slots), with slot names identical to the corrosponding block names in Popup.twig

An easy way to test this popup would be to apply [this patch](https://gist.github.com/dorian451/d9428efbf1858d569afae437c47c718f#file-10782-test-patch) and edit the resulting `test.vue` file.

WIP: Add pictures

Only defaults: 2024-07-30T14:20:22,161490105-04:00

### CODE REVIEW

File: site/app/templates/Vue.twig [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <script> 
     $('document').ready(() => {
-        window.submitty.render('[data-vuemount]', "{{ page }}",  {{ args | json_encode(constant("JSON_FORCE_OBJECT")) | raw() }});
```
Comment (williamjallen) [inline site/app/templates/Vue.twig:3]: Should this change be made in a separate bugfix PR?

File: site/vue/src/components/popup.vue [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            }
+        }
+    },
+    { immediate: true },
+);
+
+</script>
+
+<template>
+  <Teleport
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:51]: How hard would it be to avoid using Teleport here?  A messy DOM structure like this will be pretty confusing for new developers, and incurs a significant amount of technical debt.

File: site/vue/src/components/popup.vue [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                class="btn btn-default close-button"
+                tabindex="-1"
+                type="button"
+                @click="$emit('dismiss')"
+              >
+                Close
+              </button>
+            </div>
+          </slot>
+          <slot name="body_panel">
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:87]: Should we give developers the ability to override the body and button bar?  I think it would be better to get rid of this slot and force developers to use the `body` and `buttons_panel` slots.

File: site/vue/src/components/popup.vue [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            <div class="form-body">
+              <slot name="body">
+                Be sure to override the body slot.
+              </slot>
+              <slot name="buttons_panel">
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:92]: Is this slot necessary?  I think it would be better to force developers to put everything in a standard preformatted button bar.  If we give developers the ability to heavily customize things like this, we'll end up in the current situation with the Twig popup where every single popup is different, and it's absolutely impossible to reliably apply styling changes to all of them at once.

File: site/vue/src/components/popup.vue [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                        <a
+                          :id="closeButton.id"
+                          class="btn close-button key_to_click"
+                          :class="closeButton.classes.join(' ')"
+                          tabindex="0"
+                          @click="$emit('dismiss')"
+                        >
+                          <slot name="close_button_text">
+                            Close
+                          </slot>
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:106]: Would it be better to use a prop for this instead?

File: site/vue/src/components/popup.vue [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            <div class="form-body">
+              <slot name="body">
+                Be sure to override the body slot.
+              </slot>
+              <slot name="buttons_panel">
+                <div class="form-buttons">
+                  <div class="form-button-container">
+                    <slot name="buttons">
+                      <slot name="close_button">
+                        <a
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:97]: Should this be a button instead, or would doing that break too much existing CSS?

File: site/vue/src/components/popup.vue [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+          <slot name="body_panel">
+            <div class="form-body">
+              <slot name="body">
+                Be sure to override the body slot.
+              </slot>
+              <slot name="buttons_panel">
+                <div class="form-buttons">
+                  <div class="form-button-container">
+                    <slot name="buttons">
+                      <slot name="close_button">
```
Comment (williamjallen) [inline site/vue/src/components/popup.vue:96]: I think the button bar should always have a close button.  Should the `buttons` slot go next to a fixed close button?

File: site/vue/tsconfig.app.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
   "compilerOptions": {
     "composite": true,
     "tsBuildInfoFile": "./node_modules/.tmp/tsconfig.app.tsbuildinfo",
+    "types": ["jquery"],
```
Comment (williamjallen) [inline site/vue/tsconfig.app.json:8]: Is this needed for this PR?  If not, it would be better to do this in a separate PR which can be merged immediately.

**Review by williamjallen — CHANGES_REQUESTED** [review]
I think this is on the right track in general, but the current approach leads us directly to a configuration hell situation which will make it virtually impossible to refactor in the future.  Instead, I think it would be better to keep things as simple as possible, while being mindful to keep things generic so we don't need dozens of little implementation-specific configuration options.

File: site/app/templates/Vue.twig (hunk 1 above)
Comment (dorian451) [reply site/app/templates/Vue.twig:3]: I don't get why this change is showing up in this PR, it was merged into main with https://github.com/Submitty/Submitty/pull/10780

Comment (dorian451) [reply site/app/templates/Vue.twig]: Oh I just needed to merge in main through github (it didnt work when I did it locally?)

File: site/vue/src/components/popup.vue (hunk 2 above)
Comment (dorian451) [reply site/vue/src/components/popup.vue]: I genutively thought it would be cleaner :(. It's not nessesary, reverted in https://github.com/Submitty/Submitty/pull/10782/commits/e7c1aa4ed6c4b5070bc5419b48f6e26ec3ce61b5

**Comment by dorian451** [comment]
> current approach leads us directly to a configuration hell situation
I do agree the way this popup config is set up is really confusing and unintuitive, but it's also the way `Popup.twig` is set up (which is what I copied this configuration scheme from).  Since it seems like that isn't what we want for this, I'll remake the popup so that it's a lot more intuitive.

**Review by williamjallen — APPROVED** [review]
Code looks good.  Great job with this!
